src/xarray_regrid/methods/conservative.py

```diff
@@ -195,6 +153,36 @@ def apply_weights(
     )
 
 
+def compute_einsum_dask(da: xr.DataArray, weights: np.ndarray) -> dask.array.Array:
+    """Compute the einsum between dask data and weights, and mask NaNs if needed."""
+    new_data: dask.array.Array
+    if np.any(np.isnan(da.data)):
+        new_data = dask.array.einsum(
+            "i...,ij->j...", da.fillna(0).data, weights, optimize="greedy"
```

Comment:
FYI. If `einsum` is the bottleneck, then there are many optimization approaches available, apart from `greedy` and `optimal`, for instance, check [`opt_einsum`](https://optimized-einsum.readthedocs.io/en/stable/).